cirq/ops/phased_iswap_gate.py

```diff
@@ -0,0 +1,197 @@
+         [0, 0, 0, 1]]
+
+    where:
+
+        c = cos(π·t/2)
+        s = sin(π·t/2)
+        f = exp(2πi·p)
+
+    and star indicates complex conjugate.
+    """
+
+    def __init__(self,
+                 *,
+                 phase_exponent: Union[float, sympy.Symbol] = 1 / 4,
+                 exponent: Union[float, sympy.Symbol] = 1.0,
+                 global_shift: float = 0.0) -> None:
+        """
+        Args:
+            phase_exponent: The exponent on the Z gates.
+            exponent: The exponent on the ISWAP gate, see EigenGate for
+                details.
+            global_shift: How much to shift the operation's eigenvalues at
+                exponent=1, see EigenGate for details.
+        """
+        self._phase_exponent = value.canonicalize_half_turns(phase_exponent)
+        super().__init__(exponent=exponent, global_shift=global_shift)
+
+    @property
+    def phase_exponent(self) -> Union[float, sympy.Symbol]:
+        return self._phase_exponent
+
+    def _json_dict_(self):
+        json_dict = super()._json_dict_()
+        json_dict['phase_exponent'] = self.phase_exponent
+        return json_dict
+
+    def _value_equality_values_cls_(self):
+        if self.phase_exponent == 0:
+            return common_gates.ISwapPowGate
+        return PhasedISwapPowGate
+
+    def _value_equality_values_(self):
+        if self.phase_exponent == 0:
+            return super()._value_equality_values_()
+        return (self.phase_exponent, *super()._value_equality_values_())
+
+    def _is_parameterized_(self) -> bool:
+        if protocols.is_parameterized(super()):
+            return True
+        return protocols.is_parameterized(self._phase_exponent)
+
+    def _with_exponent(self,
+                       exponent: type_alias.TParamVal) -> 'PhasedISwapPowGate':
+        return PhasedISwapPowGate(phase_exponent=self.phase_exponent,
+                                  exponent=exponent,
+                                  global_shift=self.global_shift)
+
+    def _eigen_shifts(self) -> List[float]:
+        return [0.0, +0.5, -0.5]
+
+    def _eigen_components(self):
+        phase = np.exp(1j * np.pi * self.phase_exponent)
+        phase_matrix = np.diag([1, phase, phase.conjugate(), 1])
+        inverse_phase_matrix = np.conjugate(phase_matrix)
+        eigen_components: List[Tuple[float, np.ndarray]] = []
+        for eigenvalue, projector in super()._eigen_components():
+            new_projector = phase_matrix @ projector @ inverse_phase_matrix
+            eigen_components.append((eigenvalue, new_projector))
+        return eigen_components
+
+    def _apply_unitary_(self, args: 'protocols.ApplyUnitaryArgs'
+                       ) -> Optional[np.ndarray]:
+        if protocols.is_parameterized(self):
+            return NotImplemented
+        if self.exponent != 1:
+            return NotImplemented
+
+        phase = np.exp(1j * np.pi * self.phase_exponent)
+        zo = args.subspace_index(0b01)
```

Comment:
You can use `cirq.linalg.apply_matrix_to_slices` to operate on the 2x2 inner block. It will be simpler and faster.

Reply:
Unfortunately, `apply_matrix_to_slices` cannot apply the matrix in-place (and in our case, we're just multiplying in some phase into each row and column - something easily done in place as it is done currently).

Reply:
Current code (arbitrary exponent): 3.9s
Current code (exponent=1): 0.47s
Apply to slices code (arbitrary exponent): 0.41s to 0.48s

New code:

```python
    def _apply_unitary_(self, args: 'protocols.ApplyUnitaryArgs'
                       ) -> Optional[np.ndarray]:
        if protocols.is_parameterized(self):
            return NotImplemented

        c = np.cos(np.pi * self._exponent / 2)
        s = np.sin(np.pi * self._exponent / 2)
        f = np.exp(2j * np.pi * self._phase_exponent)
        matrix = np.array([
            [c, 1j*s*f],
            [1j*s*f.conjugate(), c]
        ])

        zo = args.subspace_index(0b01)
        oz = args.subspace_index(0b10)
        linalg.apply_matrix_to_slices(
            args.target_tensor,
            matrix,
            [oz, zo],
            out=args.available_buffer)
        return args.available_buffer
```

Test case:

```python
import cirq

qubits = cirq.LineQubit.range(18)
c = cirq.Circuit(
    cirq.PhasedISwapPowGate(exponent=0.1432, phase_exponent=0.6542).on(a, b)
    for a in qubits
    for b in qubits
    if a != b
)

import time
t0 = time.monotonic()
cirq.final_wavefunction(c)
t1 = time.monotonic()
print(t1 - t0)
```

Reply:
Very cool! Done.